File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        long long int prod = 1;
        bool foundZero = false;
        int countZero = 0;
        for(int num : nums)
        {
            if(num == 0)
            {
                foundZero = true;
                countZero++;
                continue;
            }
            prod = prod * num;
        }
        vector<int> ans;
        int n = nums.size();
        for(int i = 0; i < n; i++)
        {
            if(nums[i] == 0 && countZero == 1)
            {
                ans.push_back(prod);
                continue;
            }
            if(foundZero)
                ans.push_back(0);
            else
                ans.push_back(prod / nums[i]);
        }

        return ans;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums)
        left = [1] * n
        right = [1] * n
        
        prev = 1
        for i in range(0, n - 1):
            left[i + 1] = prev * nums[i]
            prev = left[i + 1]
        
        prev = 1
        for i in range(n - 2, -1, -1):
            right[i] = prev * nums[i + 1]
            prev = right[i]
        
        ans = []
        for i in range(n):
            ans.append(left[i] * right[i])
        
        return ans